### PR TITLE
Move from static to dynamic libraries.

### PR DESCRIPTION
**Pull request by guacke**
This is the first in a series of commits where we move from static to dynamic libraries. As numopt and gqcp are based on this library, this can cause (temporary) Travis errors in those development branches, until these libraries have also been update to dynamic.

### CODE REVIEW

File: cmake/ConfigureLibrary.cmake [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
-target_include_directories(${LIBRARY_NAME} PRIVATE ${EIGEN3_INCLUDE_DIR})
+target_link_libraries(${LIBRARY_NAME} PUBLIC Eigen3::Eigen)
 
 # Include Boost
-target_include_directories(${LIBRARY_NAME} PRIVATE ${Boost_INCLUDE_DIRS})
+target_include_directories(${LIBRARY_NAME} PUBLIC ${Boost_INCLUDE_DIRS})
+target_link_libraries(${LIBRARY_NAME} PUBLIC ${Boost_LIBRARIES})
+
+# Position independent code
+set_property(TARGET ${LIBRARY_NAME} PROPERTY POSITION_INDEPENDENT_CODE ON)
```
Comment (lelemmen) [inline cmake/ConfigureLibrary.cmake]: Add newline at EOF

File: cmake/ConfigureLibrary.cmake [hunk 2]
```diff
@@ -1,11 +1,16 @@
 # In this CMake file, we will include the headers and link to the necessary libraries
 
-
 # Include this project's headers
-target_include_directories(${LIBRARY_NAME} PRIVATE ${PROJECT_INCLUDE_FOLDER})
+target_include_directories(${LIBRARY_NAME} PUBLIC
+        $<BUILD_INTERFACE:${PROJECT_INCLUDE_FOLDER}>
+        $<INSTALL_INTERFACE:${PROJECT_INSTALL_INCLUDE_FOLDER}>)
```
Comment (lelemmen) [inline cmake/ConfigureLibrary.cmake]: Using these generator expressions, during the CMake 'build', the include files are searched in `${PROJECT_INCLUDE_FOLDER}`, and when the project is installed, they are searched in `${PROJECT_INSTALL_INCLUDE_FOLDER}`, right?

File: cmake/ConfigureLibrary.cmake (hunk 1 above)
Comment (lelemmen) [inline cmake/ConfigureLibrary.cmake]: According to [this page](https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html), `POSITION_INDEPENDENT_CODE` is `ON` by default if the target is a shared library.

File: cmake/ConfigureTest.cmake [hunk 3]
```diff
@@ -1,12 +1,10 @@
 # In this CMake file, we will include header files and link to libraries for a given test source
 
 # Include this project
-target_include_directories(${TEST_NAME} PRIVATE ${PROJECT_INCLUDE_FOLDER})
-target_link_libraries(${TEST_NAME} PRIVATE ${LIBRARY_NAME} rt)
+# See: https://stackoverflow.com/questions/25676277/cmake-target-include-directories-prints-an-error-when-i-try-to-add-the-source
```
Comment (lelemmen) [inline cmake/ConfigureTest.cmake]: Maybe you could add this reference to the `ConfigureLibrary.cmake`-file too?

File: cmake/SetCMakeEnvironment.cmake [hunk 4]
```diff
@@ -1,26 +1,28 @@
 # In this CMake file, all CMake variables will be set
 
-
 # Parse the project name into uppercase and lowercase
 string(TOUPPER ${PROJECT_NAME} PROJECT_NAME_UPPERCASE)  # ppercase is needed in version.hpp.in
```
Comment (lelemmen) [inline cmake/SetCMakeEnvironment.cmake]:
```suggestion
string(TOUPPER ${PROJECT_NAME} PROJECT_NAME_UPPERCASE)  # uppercase is needed in version.hpp.in
```

**Comment by guacke** [comment]
This commit is needed for GQCG/gqcp#153.

**Comment by lelemmen** [comment]
I'm closing this PR because this library is merged into gqcp.

**Pull request closed without merging**